### ISSUE TITLE
Developer experience feedback (cli, setup, tooling, etc.)

### ISSUE THREAD

**Issue by fubhy** [streamingfast/substreams#122]
Oy! I'm going to keep adding unfiltered, brutally honest feedback to this issue. I'll try to keep it short for each individual bullet point. For now, these are all about the developer experience of actually writing substreams modules (map/store), project setup, cli, etc.

I'll open a separate issue for Rust module implementation specific developer experience feedback (developer experience of actually writing module code).

So, consider this (and the other issue I'm about to create) my unfiltered feedback notepad. Comments welcome.

Command line interface
---
 - Commands like `substreams run`, `substreams graph`, etc. should default to using the cwd's `substreams.yaml`
 - `substreams run` in `--ui` mode should not tail log output for every single block and instead provide some kind of ui that tracks progress separately from explicit output. For a lot of less active substream modules, relevant logs & data are otherwise completely swallowed in a flood of irrelevant output.
 - `substreams run` could provide an interactive mode that lets you incrementally fetch batches of blocks instead of just finishing after `--stop-block`. For instance, this could also enable mode that `breaks` on the first received non-nil output and then lets you continue until the next is received, etc.
 - `substreams codegen` seems to be misplaced in that cli. Instead, let's just make the `buf` related configuration part of the default template repository & scaffolding and afterwards leave it to the user.

Code generation
---
- The fact that the entire project needs to be in a errorless state & compile before anything in `build.rs` can run is rather cumbersome in combination with code generation via `buf`, `abigen`, etc. when refactoring or moving stuff around. Not sure what'd be the best way to solve this better but I ran into this several times where, during refactoring, I ended up in a checken-and-egg situation with generated code being out of sync with my changes but not being able to fix it with codegen because `build.rs` wouldn't run.

Pre-flight checks
---
- Typos or mismatches in e.g. `substreams.yaml` store/map configuration (e.g. store names, source names, proto message names, etc.) vs. their counterparts in the lib code are not pointed out clearly. There should be extensive pre-flight checks that validate the `substreams.yaml` configuration against the handler code in the crate.
- This can likely be done by extending the macro code further to validate `substreams.yaml` with the reality of the module code much more than it currently does.
- To name a few important checks:
	- Check if module input & output type match with substreams.yaml declarations
	- Check for initial block clashes (e.g. impossible combinations)
	- Check for circular dependencies
	- Check for missing store / map modules (defined but not implemented or vica versa)
	- Check for stores without a dependent module (likely means that a store is incorrectly used as a sink)
	- Check for missing proto messages

Editor integration
---
- The `substreams.yaml` should have a JSON Schema (https://json-schema.org/) published on https://www.schemastore.org/json/ and also uploaded & versioned somewhere for use in annotations. This would give IDEs the ability to automatically validate & lint (e.g. point out typos, help with allowed enum values, etc.) the YAML file.

JetBrains: https://www.jetbrains.com/help/idea/json.html#ws_json_schema_store_enable_auto_download)
VSCode: https://github.com/redhat-developer/vscode-yaml

In both cases, schemastore is supported by default.

**Comment by maoueh** [comment]
Lot's of great ideas that indeed we should implement for the next version. I agree to most point, here some thoughts on some of them:

> substreams run in --ui mode should not tail log output for every single block and instead provide some kind of ui that tracks progress separately from explicit output. For a lot of less active substream modules, relevant logs & data are otherwise completely swallowed in a flood of irrelevant output.

and

> substreams run could provide an interactive mode that lets you incrementally fetch batches of blocks instead of just finishing after --stop-block. For instance, this could also enable mode that breaks on the first received non-nil output and then lets you continue until the next is received, etc.

We want to improve the UI experience of the tool. We had good debates internally about how the tool should look like. For should it be web based (CLI opens locally browser) or TUI based. Some of the changes in upcoming release are a starting point for a better debugging experience like all module's output in development mode.

> substreams codegen seems to be misplaced in that cli. Instead, let's just make the buf related configuration part of the default template repository & scaffolding and afterwards leave it to the user. I think we should leave it to the user to choose their preferred protoc code generation mechanic. Buf is really just an implementation detail for that part.

Not against, I see the point of making explicit `buf`. Would like to see others opinion.

> The fact that the entire project needs to be in a errorless state & compile before anything in build.rs can run is rather cumbersome in combination with code generation via buf, abigen, etc. when refactoring or moving stuff around. Not sure what'd be the best way to solve this better but I ran into this several times where, during refactoring, I ended up in a checken-and-egg situation with generated code being out of sync with my changes but not being able to fix it with codegen because build.rs wouldn't run.

I think we fix that, would need to check. I know we have internals to remove some validation check. Is it you @Eduard-Voiculescu that did something around that?

> Pre-flight checks

Agreed. I don't think we can easily check the actual types, parameters in WASM being pointer to memory location, I don't think the types are easily available (maybe not at all).

> Managing smart contract artifacts

It's a good idea. I would see not the CLI drive that because it's a chain specific thing, so IMO, `substreams-ethereum` via a `build.rs` that reads this `abis.yaml`. 

I recently though that we should add `mod.rs` auto-generation also. I've tried `protoc-gen-prost-crate` which does generate the aggregate `mod.rs` of all Protobufs. It's cool to not have to manage `mod.rs` also.

I'll try to come up with some concrete planning of all this.

**Comment by abourget** [comment]
`substreams codegen` merely calls `buf` if there's a `buf.gen.yaml` file in there. It won't touch it. But I agree we could split the concerns.

**Comment by maoueh** [comment]
> The fact that the entire project needs to be in a errorless state & compile before anything in build.rs can run is rather cumbersome in combination with code generation via buf, abigen, etc. when refactoring or moving stuff around. Not sure what'd be the best way to solve this better but I ran into this several times where, during refactoring, I ended up in a checken-and-egg situation with generated code being out of sync with my changes but not being able to fix it with codegen because build.rs wouldn't run.

@fubhy Seems to be fixed in latest version, can you confirm? I tested here on a project that had errors, was able to generate the Proto.

**Comment by maoueh** [comment]
Not ideal integration, but not that 100% bad either. Maybe some hybrid, not sure.

**Comment by fubhy** [comment]
Yeah that's what I meant with externalizing it. If that works well then that'd definitely be better than this chicken-and-egg situation that we have atm. :)

**Comment by abourget** [comment]
Broken down into several other issues.
